typing/ctype.ml

```diff
@@ -2371,7 +2371,7 @@ and mcomp_type_decl type_pairs env p1 p2 tl1 tl2 =
       | Type_record (lst,r), Type_record (lst',r') when r = r' ->
           mcomp_list type_pairs env tl1 tl2;
           mcomp_record_description type_pairs env lst lst'
-      | Type_variant v1, Type_variant v2 ->
+      | Type_variant (v1,r), Type_variant (v2,r') when r = r' ->
```

Comment:
This change looks like it may catch a bug. Does it?

Reply:
It is a mild improvement of behaviour, but I would not go so far as to call the old behaviour buggy.

`mcomp` is used during GADT matches to determine that two types cannot possibly be equal (even via internal definitions not currently visible). It can already determine that two variant types with different constructor names or arities cannot be equal, and now it can also determine that two variant types that differ only in unboxiness cannot be equal.

This makes it marginally better at refuting impossible cases, but the old behaviour was also sound.